DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,194 @@
+- [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) - "Don't repeat yourself" suggests you should preference code re-use over duplication. However, you should [avoid hasty abstractions](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction) as the wrong abstraction can be extremely costly to correct. Instead, lean into small amounts of duplication to help identify the right abstractions through multiple use cases and define the pathway to DRY design.
+
+## "clean code that is optimized for change"
+
+What constitutes clean code? We want simple and maintainable, but how do we define that? The following best practices offer niche advice. Combined, they represent our definition of clean code.
+
+### State
+
+> Shared mutable state is believed by many to be the “root of all evil”, or at least the cause of most of the accidental complexity in our code. And “Complexity is the root cause of the vast majority of problems with software today.” - [Mauro Bieg](https://mb21.github.io/blog/2021/01/23/pure-functional-programming-and-shared-mutable-state.html#the-root-of-all-evil)
+
+Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.
+
+In avoiding package level state we avoid [hidden dependencies and unintended side effects](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state) stemming from global state mutation. Model loosely-coupled components by [declaring your dependencies](#dependencies).
+
+### Names
+
+##### Packages
+
+Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. 
+
+Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Decorator`. You may find you have a type named after the package, soemtimes it's appropriate.
+
+If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.
+
+Avoid (like the plague) `types`, `interfaces`, `common`, `util` or `base` packages. They don't represent anything cohesive and tend to collect unrelated types and algorithms that often require teething apart as a project progresses due to cyclic dependencies. If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.
+
+##### Function & method
+
+Functions and methods (functions with a receiver parameter) should adequetly describe their behavior. Generally, they should follow a verb-noun form. The types of parameters should help inform the function name.
```

Comment:
Parameter types help inform on the functions scope/purpose, not necessarily its name.